pkg/ds/farm.go

```diff
@@ -0,0 +1,411 @@
+
+	logger  logging.Logger
+	alerter alerting.Alerter
+}
+
+type childDS struct {
+	ds        DaemonSet
+	quitCh    chan<- struct{}
+	updatedCh chan<- *ds_fields.DaemonSet
+	deletedCh chan<- *ds_fields.DaemonSet
+	errCh     <-chan error
+}
+
+func NewFarm(
+	kpStore kp.Store,
+	dsStore dsstore.Store,
+	applicator labels.Applicator,
+	logger logging.Logger,
+	alerter alerting.Alerter,
+) *Farm {
+	if alerter == nil {
+		alerter = alerting.NewNop()
+	}
+
+	return &Farm{
+		kpStore:    kpStore,
+		dsStore:    dsStore,
+		scheduler:  scheduler.NewApplicatorScheduler(applicator),
+		applicator: applicator,
+		children:   make(map[fields.ID]*childDS),
+		logger:     logger,
+		alerter:    alerter,
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			if err != nil {
+				dsf.logger.NoFields().Errorf("Unable to get daemon sets from intent tree: %v", err)
+				continue
+			}
+
+			dsIDMap := make(map[fields.ID]ds_fields.DaemonSet)
+			for _, dsFields := range allDaemonSets {
+				dsIDMap[dsFields.ID] = dsFields
+			}
+
+			dsIDLabelSelector := klabels.Everything().
+				Add(DSIDLabel, klabels.ExistsOperator, []string{})
+
+			allPods, err := dsf.applicator.GetMatches(dsIDLabelSelector, labels.POD)
+			if err != nil {
+				dsf.logger.NoFields().Error(err)
+				continue
+			}
+
+			for _, podLabels := range allPods {
+				// Only check if it is a pod scheduled by a daemon set
+				dsID := podLabels.Labels.Get(DSIDLabel)
+
+				// Check if the daemon set exists, if it doesn't unschedule the pod
+				if _, ok := dsIDMap[fields.ID(dsID)]; !ok {
+					nodeName, podID, err := labels.NodeAndPodIDFromPodLabel(podLabels)
+					if err != nil {
+						dsf.logger.NoFields().Error(err)
+						continue
+					}
+
+					// TODO: Since this mirrors the unschedule function in daemon_set.go,
+					// We should find a nice way to couple them together
+					dsf.logger.NoFields().Infof("Unscheduling from '%s' with dangling daemon set uuid '%s'", nodeName, dsID)
+
+					_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+					if err != nil {
+						dsf.logger.NoFields().Errorf("Unable to delete pod id '%v' from intent tree: %v", dsID, err)
+
+						if alertErr := dsf.alerter.Alert(alerting.AlertInfo{
+							Description: fmt.Sprintf("Unable to delete the following pod from store: %v", podID),
+							IncidentKey: "pod_store_delete_error",
```

Comment:
now that i think about it i'm not sure we want to page people if a pod fails to delete though, especially since it gets retried